### PR TITLE
Statstest from Colorado School of Mines

### PR DESCRIPTION
**Pull request by FlanFlanagan**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
No issue tied to this PR

##### What are the significant changes in functionality due to this change request?
This PR adds an additional metric scheme to test that validity of synthetic data. It currently includes Chi_square, Kolmogorov–Smirnov test, and F-tests.  Additionally it allows for data to be sorted in a number of ways. By segment length, by interval (i.e. every x'th data point), or by summation of each segment. 

I will note, I'm not sure what the changes on the second two files are from. I'm up to date with idaholab/raven/devel. I'm happy to adjust those files to match the current files!

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Metrics/metrics/StatsTestMetric.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  available_tests['full']['ks_test'] = stats.ks_2samp
+  available_tests['seg']['f_test'] = sU.compare_segments_f
+  available_tests['seg']['chi_square'] = sU.compare_segments_chi
+  available_tests['seg']['ks_test'] = sU.compare_segments_ks
+  available_tests['int']['f_test'] = sU.compare_interval_f
+  available_tests['int']['chi_square'] = sU.compare_interval_chi
+  available_tests['int']['ks_test'] = sU.compare_interval_ks
+  available_tests['seg_sum']['f_test'] = sU.compare_sums_f
+  available_tests['seg_sum']['chi_square'] = sU.compare_sums_chi
+  available_tests['seg_sum']['ks_test'] = sU.compare_sums_ks
```
Comment (PaulTalbot-INL) [inline framework/Metrics/metrics/StatsTestMetric.py:43]: RAVEN naming conventions don't allow for `snake_case`, requiring all variables to be `camelCase`.

File: framework/Metrics/metrics/StatsTestMetric.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        or given distribution.
+      @ In, y, instance of Distributions.Distribution, tuple or list, array containing data of y,
+        or given distribution.
+      @ In, weights, array_like (numpy.ndarray or list), optional, not used in this metric
+      @ In, axis, integer, optional, default is 0, not used for this metric.
+      @ In, kwargs,dict, dictionary of parameters characteristic of each metric
+      @ Out, value, float, metric result, CDF area difference
+    """
+    if isinstance(x,np.ndarray) and isinstance(y,np.ndarray):
+      assert x.shape == y.shape, "Input data x, y should have the same shape!"
```
Comment (PaulTalbot-INL) [inline framework/Metrics/metrics/StatsTestMetric.py:85]: Note that asserts are skipped when run in default RAVEN mode (python's opt mode). Should this be a `self.raiseAnError` or is this assertion just a developer check?

File: framework/Metrics/metrics/StatsTestsUtils.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+from scipy import stats
+
+
+def build_segments(array, seg_length: int):
+"""
+This method groups elements of an array into segments of length seg_length
+@In array, array-like containing data to be grouped into segments.
+@In seg_length, size of the segment desired
+@Out, array of segments of length seg_length
+"""
```
Comment (PaulTalbot-INL) [inline framework/Metrics/metrics/StatsTestsUtils.py:12]: docstring open/closer should be indented once, and the docstring contents indented twice, as:
```python
def func(args):
  """
    description
    @ In, args, args type, description
    (etc)
  """
  codeLine1 = True
```
then the code should be indented once as well, with 2-space indents everywhere

File: framework/Metrics/metrics/StatsTestsUtils.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
+import numpy as np
+
+from scipy import stats
+
+
+def build_segments(array, seg_length: int):
```
Comment (PaulTalbot-INL) [inline framework/Metrics/metrics/StatsTestsUtils.py:6]: snake case is not allowed for methods as well, they should be camelCase

File: framework/SupervisedLearning/ARMA.py [hunk 5]
```diff
@@ -278,7 +278,7 @@ def __init__(self):
     self.zeroFilterTarget  = None # target for whom zeros should be filtered out
     self.zeroFilterTol     = None # tolerance for zerofiltering to be considered zero, set below
     self._masks            = collections.defaultdict(dict)   # dictionay of masks, including zeroFilterMask(where zero), notZeroFilterMask(Where non zero), and maskPeakRes.
-    self._minBins          = 20   # min number of bins to use in determining distributions, eventually can be user option, for now developer's pick
+    self._minBins          = 50   # min number of bins to use in determining distributions, eventually can be user option, for now developer's pick
```
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/ARMA.py:281]: I think you'll want to revert this change, otherwise we'll have to regold all the ARMA regression tests.

File: framework/SupervisedLearning/ARMA.py [hunk 6]
```diff
@@ -616,7 +616,7 @@ def __trainLocal__(self, featureVals, targetVals):
     # Applied Energy, 87(2010) 843-855
     for t,target in enumerate(self.target):
       # if target correlated with the zero-filter target, truncate the training material now?
-      timeSeriesData = targetVals[:,t]
+      # timeSeriesData = targetVals[:,t]
```
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/ARMA.py:619]: why this change? For multiple targets I would imagine this is valid.

File: framework/SupervisedLearning/ARMA.py [hunk 7]
```diff
@@ -2009,6 +2008,8 @@ def getGlobalRomSegmentSettings(self, trainingDict, divisions):
       pivotValues = trainingDict[self.pivotParameterID][0]
       # use the first segment as typical of all of them, NOTE might be bad assumption
       delta = pivotValues[slicers[0][-1]] - pivotValues[slicers[0][0]]
+      # # NOTE: if moving the segmentation length Fourier after segmentation uncomment the line below
+      delta = pivotValues[slicers[1][0]] - pivotValues[slicers[0][0]]
```
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/ARMA.py:2012]: what does this mean? The comment is helpful, but not quite clear yet to me

File: framework/SupervisedLearning/ARMA.py [hunk 8]
```diff
@@ -2020,15 +2021,19 @@ def getGlobalRomSegmentSettings(self, trainingDict, divisions):
...
+          #segment[target] = periods[np.logical_not(periods > delta)]
+          # if len(full):
+          #   # train Fourier on longer periods
+          #   self.fourierResults[target] = self._trainFourier(pivotValues, full, targetVals, target=target)
+          #   # self.fourierResults[target] = self._trainFourier(pivotValues, full, targetVals, masks=self._masks[self.zeroFilterTarget]['notZeroFilterMask'], target=target)
+          #   # remove longer signal from training data
+          #   signal = self.fourierResults[target]['predict']
+          #   targetVals = np.array(targetVals, dtype=np.float64)
+          #   targetVals -= signal
+          #   trainingDict[target][0] = targetVals
```
Comment (PaulTalbot-INL) [inline framework/SupervisedLearning/ARMA.py:2036]: it looks to me like this disables handling any Fourier fitting of the full signal before segmenting. Is that right?

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
Sorry it took me so long to get to this.

There's some style guide issues that need addressing, especially in the utils file, and a couple questions about changes to the ARMA to consider. Otherwise, this looks very useful so far.

Are there any regressions tests to assure the metrics work as expected, and demonstrate how they might be used?

**Comment by wangcj05** [comment]
@FlanFlanagan @PaulTalbot-INL Do you have a plan to resolve all the comments and finalize this PR?

**Comment by wangcj05** [comment]
@dylanjm If we do squash and merge, I think we can avoid the break of git history.

**Comment by FlanFlanagan** [comment]
@wangcj05 I do! Sorry, I'm back logged and should get to it this weekend. 

I'm happy to do what needs to be done to make sure it doesn't break history. I'm not sure where the moosetest account stuff is coming from.

File: framework/SupervisedLearning/ARMA.py (hunk 8 above)
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/ARMA.py:2036]: I believe some of these changes may have been necessary to get the fits you were using for the solar paper, that may have come from a branch of Andrea that never got merged to `devel`. If you want to close this and rebase, I would still maintain this PR and associated branch as a list of changes that "was working" in case you run into issues in the rebase.

Comment (wangcj05) [reply framework/SupervisedLearning/ARMA.py:2036]: Create a new one is good to me. It seems this PR gets some weird behaviors.

Comment (FlanFlanagan) [reply framework/SupervisedLearning/ARMA.py:2036]: Ahh that makes sense Paul. I'll do that. Thanks!

**Comment by wangcj05** [comment]
> @wangcj05 @dylanjm @PaulTalbot-INL The files modified outside the StatsTestMetric and Utils files are not changes I made. They must be relics of an old branch. Would you rather I close this PR and open a new PR based off the most recent dev branch with just me new files added?
> 
> Thanks!

@FlanFlanagan: Any update on this PR? Have you created a new PR on this development?

**Comment by FlanFlanagan** [comment]
Yep! Sorry, I should have pinged here. It is PR #1788

**Comment by wangcj05** [comment]
> Yep! Sorry, I should have pinged here. It is PR #1788

Thanks for the update, this PR can be closed since a new PR is open now.
